File: Data Structures & Algorithms/surrounded-regions/submission-9.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        rows = len(board)
        cols = len(board[0])
        q = deque()
        directions = [[-1, 0], [0, -1], [1, 0], [0, 1]]

        for r in range(rows):
            if board[r][0] == 'O':
                q.append((r, 0))
            if board[r][cols - 1] == 'O':
                q.append((r, cols - 1))

        for c in range(1, cols - 1):
            if board[0][c] == 'O':
                q.append((0, c))
            if board[rows - 1][c] == 'O':
                q.append((rows - 1, c))

        while q:
            r, c = q.popleft()

            if board[r][c] == 'O':
                board[r][c] = 'C'

                for dr, dc in directions:
                    nr = r + dr
                    nc = c + dc
                    if 0 <= nr < rows and 0 <= nc < cols:
                        q.append((nr, nc))

        for r in range(rows):
            for c in range(cols):
                if board[r][c] == 'O':
                    board[r][c] = 'X'
                elif board[r][c] == 'C':
                    board[r][c] = 'O'

File: Data Structures & Algorithms/rotting-fruit/submission-9.ts
class Solution {
    /**
     * @param {number[][]} grid
     * @return {number}
     */
    orangesRotting(grid: number[][]): number {
        const ROWS = grid.length
        const COLS = grid[0].length
        let fresh = 0
        let sources: [number, number][] = []

        for (let r = 0; r < ROWS; r++) {
            for (let c = 0; c < COLS; c++) {
                if (grid[r][c] === 1) {
                    fresh++
                } else if (grid[r][c] === 2) {
                    sources.push([r, c])
                }
            }
        }

        if (fresh === 0) return 0

        let head = 0
        let time = 0
        const directions = [[1, 0], [0, 1], [-1, 0], [0, -1]]
        
        while (head < sources.length) {
            let n = sources.length
            for (let i = head; i < n; i++){
                const [r, c] = sources[head++]

                for (const [dr, dc] of directions) {
                    const newRow = r + dr
                    const newCol = c + dc
                    if (newRow >= 0 && newCol >= 0 && newRow < ROWS && newCol < COLS && grid[newRow][newCol] === 1) {
                        grid[newRow][newCol] = 2
                        fresh--
                        sources.push([newRow, newCol])
                    }
                }

                if (fresh === 0) {
                    return time + 1
                }
            }
            time++
        }
        return -1
    }
}
